src/test/librbd/test_Groups.cc

```diff
@@ -480,3 +480,262 @@ TEST_F(TestGroup, add_snapshotPP)
+  ASSERT_EQ(0, rbd_group_snap_list2(ioctx, gp_name, gp_snaps, &num_snaps));
+  ASSERT_EQ(0U, num_snaps);
+
+  const char* const gp_snap_names[] = {
+    "snap_snapshot0", "snap_snapshot1", "snap_snapshot2", "snap_snapshot3"};
+  ASSERT_EQ(0, rbd_group_snap_create(ioctx, gp_name, gp_snap_names[0]));
+
+  ASSERT_EQ(0, rbd_group_image_add(ioctx, gp_name, ioctx,
+                                   m_image_name.c_str()));
+  ASSERT_EQ(0, rbd_group_snap_create(ioctx, gp_name, gp_snap_names[1]));
+
+  ASSERT_EQ(0, rbd_group_image_add(ioctx, gp_name, ioctx2,
+                                   image_name2.c_str()));
+  ASSERT_EQ(0, rbd_group_snap_create(ioctx, gp_name, gp_snap_names[2]));
+
+  ASSERT_EQ(0, rbd_group_image_remove(ioctx, gp_name, ioctx,
+                                      m_image_name.c_str()));
+  ASSERT_EQ(0, rbd_group_snap_create(ioctx, gp_name, gp_snap_names[3]));
+
+  num_snaps = 3U;
+  ASSERT_EQ(-ERANGE, rbd_group_snap_list2(ioctx, gp_name, gp_snaps,
+                                          &num_snaps));
+  ASSERT_EQ(4U, num_snaps);
+  ASSERT_EQ(0, rbd_group_snap_list2(ioctx, gp_name, gp_snaps, &num_snaps));
+
+  for (int i = 0; i < 4; i++) {
+    if (!strcmp(gp_snaps[i].name, gp_snap_names[0])) {
+      ASSERT_EQ(0U, gp_snaps[i].image_snaps_count);
+    } else if (!strcmp(gp_snaps[i].name, gp_snap_names[1])) {
+      ASSERT_EQ(1U, gp_snaps[i].image_snaps_count);
+      ASSERT_STREQ(m_image_name.c_str(),
+                   gp_snaps[i].image_snaps[0].image_name);
+      ASSERT_EQ(rados_ioctx_get_id(ioctx), gp_snaps[i].image_snaps[0].pool_id);
+    } else if (!strcmp(gp_snaps[i].name, gp_snap_names[2])) {
+      ASSERT_EQ(2U, gp_snaps[i].image_snaps_count);
+      for (int j = 0; j < 2; j++) {
+	if (!strcmp(gp_snaps[i].image_snaps[j].image_name,
+                    m_image_name.c_str())) {
+	  ASSERT_EQ(rados_ioctx_get_id(ioctx),
+                    gp_snaps[i].image_snaps[j].pool_id);
+	} else if (!strcmp(gp_snaps[i].image_snaps[j].image_name,
+                           image_name2.c_str())) {
+	  ASSERT_EQ(rados_ioctx_get_id(ioctx2),
+                    gp_snaps[i].image_snaps[j].pool_id);
+	} else {
+          FAIL() << "Unexpected image in group snap: "
+                 << gp_snaps[i].image_snaps[j].image_name;
+	}
+      }
+    } else if (!strcmp(gp_snaps[i].name, gp_snap_names[3])) {
+      ASSERT_EQ(1U, gp_snaps[i].image_snaps_count);
+      ASSERT_STREQ(image_name2.c_str(), gp_snaps[i].image_snaps[0].image_name);
+      ASSERT_EQ(rados_ioctx_get_id(ioctx2),
+                gp_snaps[i].image_snaps[0].pool_id);
+    } else {
+      FAIL() << "Unexpected group snap: " << gp_snaps[i].name;
+    }
+  }
+
+  for (const auto& gp_snap_name : gp_snap_names) {
+    ASSERT_EQ(0, rbd_group_snap_remove(ioctx, gp_name, gp_snap_name));
+  }
+  rbd_group_snap_list2_cleanup(gp_snaps, num_snaps);
+  free(gp_snaps);
+  ASSERT_EQ(0, rbd_group_snap_list2(ioctx, gp_name, gp_snaps, &num_snaps));
```

Comment:
Passing NULL would be better since `gp_snaps` just got freed:
```suggestion
  ASSERT_EQ(0, rbd_group_snap_list2(ioctx, gp_name, NULL, &num_snaps));
```